### PR TITLE
feat: DHIS2-10218 part 1 - User can see a list of the different stages

### PR DESCRIPTION
**Pull request by JoakimSM**
[DHIS2-10218](https://jira.dhis2.org/browse/DHIS2-10218)

First and main part of this ticket. Adds a generic widget component (that is being used in the stages and events widget) and lists the program stages for a given program.

### CODE REVIEW

File: src/core_modules/capture-ui/IconButton/IconButton.component.js
Comment (JoakimSM) [inline src/core_modules/capture-ui/IconButton/IconButton.component.js:1]: We will need replace the old material-UI IconButton that is being used with this one (or at least this is a start, might need to expand on this to fulfill the requirements). Let's do that as part of the already existing ticket we have.

File: cypress/integration/LockedSelector.feature
Comment (JoakimSM) [inline cypress/integration/LockedSelector.feature:141]: Had to change the order here because all the selectors didn't always load in time.

File: src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js
Comment (JoakimSM) [inline src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js:3]: Made some changes there to ensure the page status is not set to "default" before the url is updated.

File: src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPageDefault/EnrollmentPageDefault.component.js
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPageDefault/EnrollmentPageDefault.component.js:19]: Hey, 
I see that you use `spacersNum.dp16` in another part of this PR to set the paddings and margins. Can you use the same pattern in this file? 
Thanks!

**Review by paschalidi — COMMENTED** [review]
Hey looks good overall. Made few commnets here and there, feel free to reach out in case my explanations dont make sense to you. 


Another point is the `Widgets` and `Widgets` component folders we creating. 

image

To be honest I would make a different component folder for every reusable widget I am creating. In my head widgets are components just a little bit more complicated perhaps. 

I see why you would want to have them in the same folder, but honestly that folder can grow very big, quite quickly. Plus we are already in the `components` folder. The components folder holds all the different components we have. I think its useful to be able to open this components folder and see all the different components we have. Now when you are having them nested, this overview becomes a little bit blurry. 

Anyway I think these are conventions a team needs to agree upon, I am just trying to start a discussion and explore different alternatives on the topic.

File: src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js
Comment (JoakimSM) [reply src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js:88]: This is a temporary workaround for the problem you correctly mentioned (the page status is set to default before the url is updated). My plan is to get rid of this status in the redux store (it also causes other problems that I commented on in the code), but this ties into https://jira.dhis2.org/browse/TECH-580 so I will fix it there.

File: src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js
Comment (JoakimSM) [reply src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPage.epics.js:133]: Good point here. Thanks!

File: src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPageDefault/EnrollmentPageDefault.component.js
Comment (JoakimSM) [reply src/core_modules/capture-core/components/Pages/Enrollment/EnrollmentPageDefault/EnrollmentPageDefault.component.js:19]: Excellent point. Will fix, let me know if I missed this in other places.

File: src/core_modules/capture-core/components/Widgets/StagesAndEventsWidget/Stages/Stages.component.js
Comment (JoakimSM) [reply src/core_modules/capture-core/components/Widgets/StagesAndEventsWidget/Stages/Stages.component.js:25]: Good point. Fixing.

File: src/core_modules/capture-core/components/Widgets/StagesAndEventsWidget/Stages/Stages.component.js
Comment (JoakimSM) [reply src/core_modules/capture-core/components/Widgets/StagesAndEventsWidget/Stages/Stages.component.js:13]: Fair point, was also thinking about whether this abstraction makes sense. Let's keep this for now, but I definitely see your point.

File: src/core_modules/capture-core/components/Widgets/StagesAndEventsWidget/StagesAndEventsWidget.component.js
Comment (JoakimSM) [reply src/core_modules/capture-core/components/Widgets/StagesAndEventsWidget/StagesAndEventsWidget.component.js:24]: Thanks, see above.

**Comment by JoakimSM** [comment]
Widgets folder: I see your point and we could certainly keep a flat structure in the components folder. We do have other nested folders in here though, like for example DataEntries (and more). At this point in time I would prefer keeping the Widgets folder (obviously), easier to see the Widgets at a glance. But if the rest of the group also feels strongly about flattening the structure here, I will obey @simonadomnisoru @huocha

**Review by paschalidi — APPROVED** [review]
Hey thanks for coming back to my suggestions! 

I will leave it to the group to decide about the non-flat vs flat folder structure.

**Comment by simonadomnisoru** [comment]
Hey, 
When it comes to the non-flat vs flat folder structure debate, I am in favor of a non-flat structure. I think the file name should be descriptive enough to be able to group similar components. For example `StagesAndEventsWidget` can become `WidgetStagesAndEvents` and all the new components that use Widget can have the name pattern `Widget...`.

Nonetheless, in this case, the `components` folder already contains some nested folders. Therefore a new nested `Widgets` folder is compliant with the homogeneity of the folder pattern. 

To sum up, I think both patterns are valid and can be used in this case.

**Comment by JoakimSM** [comment]
Ok, gave it some thought; I realised that DataEntries (that I used in the example above) is a bit different. The DataEntries are more closely tied together since they share som common logic that are reused. So I will suggest: Components that are closely tied together should be able to live in a nested folder structure even if they are all exposed and are reusable. Especially if they share some common logic that are reused.

Widgets will most likely not be in the above category and should be separate folders directly in components like you suggested. I will move the files to WidgetStagesAndEvent (I like the suggested folder name).

I'll do this after I've merged the two other PRs in this chain into this one to avoid potential problems with renaming and moving files around.

**Review by simonadomnisoru — APPROVED** [review]

**Pull request merged**
